snap/snap.go

```diff
@@ -49,59 +52,241 @@ func addPointsAndSnap(ix *PointIndex, polygon *geom.Polygon) *geom.Polygon {
 		}
 		switch len(newRing) {
 		case 0:
-			if ringI == 0 {
+			if ringIdx == 0 {
 				// outer ring has become too small
 				return nil
 			}
 		case 1, 2:
-			if ringI == 0 {
+			if ringIdx == 0 {
 				// keep outer ring as point or line
 				newPolygon = append(newPolygon, newRing)
 			}
 		default:
 			// deduplicate points in the ring, then add to the polygon
-			newPolygon = append(newPolygon, deduplicateRing(newRing))
+			newPolygon = append(newPolygon, kmpDeduplicate(newRing))
 		}
 	}
 	return (*geom.Polygon)(&newPolygon)
 }
 
-//nolint:cyclop,nestif
-func deduplicateRing(newRing [][2]float64) [][2]float64 {
-	dedupedRing := make([][2]float64, 0, len(newRing))
-	skip := 0
-	skipped := false
-	for newVertexI, newVertex := range newRing {
-		if skip > 0 {
-			skip--
-			skipped = true
-		} else {
-			// if vertex at i is equal to vertex at i+2, and vertex at i+1 is equal to vertex at i+3,
-			// then we're just traversing the same line three times --> skip two points to turn it into a single line
-			newVertexIplus1 := (newVertexI + 1) % len(newRing)
-			newVertexIplus2 := (newVertexI + 2) % len(newRing)
-			newVertexIplus3 := (newVertexI + 3) % len(newRing)
-			if newVertexI != newVertexIplus2 && newVertexIplus1 != newVertexIplus3 &&
-				newVertex == newRing[newVertexIplus2] && newRing[newVertexIplus1] == newRing[newVertexIplus3] {
-				skip = 2
-			}
-			// if we just skipped points, there may still be a duplicate traversal from the previous point, check that in the same manner
-			if skipped {
-				newVertexIminus1 := (newVertexI - 1) % len(newRing)
-				if newVertexIminus1 != newVertexIplus1 && newVertexI != newVertexIplus2 &&
-					newRing[newVertexIminus1] == newRing[newVertexIplus1] && newVertex == newRing[newVertexIplus2] {
-					skip = 2
+// deduplication using an implementation of the Knuth-Morris-Pratt algorithm
+//
+//nolint:cyclop,funlen,nestif
+func kmpDeduplicate(newRing [][2]float64) [][2]float64 {
+	dedupedRing := make([][2]float64, len(newRing))
+	copy(dedupedRing, newRing)
+	// map of indices to remove, sorted by starting index of each sequence to remove
+	indicesToRemove := sortedmap.New(len(newRing), func(x, y interface{}) bool {
+		return x.([2]int)[0] < y.([2]int)[0]
+	})
+	// walk through newRing until a step back is taken, then identify how many steps back are taken and search for repeats
+	visitedPoints := [][2]float64{}
+	for i := 0; i < len(newRing); {
+		vertex := newRing[i]
+		if len(visitedPoints) > 1 && visitedPoints[len(visitedPoints)-2] == vertex {
```

Comment:
bijna alles zit in deze if. mogelijk kun je hem negaten (is dat een woord?) en er een guard clause van maken met een continue. dat scheelt een nesting.

Reply:
Goeie, dat is een nettere oplossing inderdaad. Heb het omgedraaid.